### PR TITLE
Add debug log for monitoring `@require` time

### PR DESCRIPTION
**Pull request by IanButterworth**
Adds a debug log so you can monitor `@require`s that get triggered and the time spent running code

```julia
julia> ENV["JULIA_DEBUG"]="Requires"
"Requires"

julia> using Foo
┌ Debug: Requires conditionally ran code in 0.917808094 seconds: `Ratios` detected `FixedPointNumbers`
└ @ Requires ~/.julia/packages/Ratios/xLeZh/src/Ratios.jl:123
┌ Debug: Requires conditionally ran code in 0.531145239 seconds: `ArrayInterface` detected `SuiteSparse`
└ @ Requires ~/.julia/packages/ArrayInterface/mI7ab/src/ArrayInterface.jl:675
┌ Debug: Requires conditionally ran code in 0.002734456 seconds: `ArrayInterface` detected `Adapt`
└ @ Requires ~/.julia/packages/ArrayInterface/mI7ab/src/ArrayInterface.jl:760
┌ Debug: Requires conditionally ran code in 0.029623663 seconds: `ArrayInterface` detected `StaticArrays`
└ @ Requires ~/.julia/packages/ArrayInterface/mI7ab/src/ArrayInterface.jl:690
┌ Debug: Requires conditionally ran code in 0.006599821 seconds: `ArrayInterface` detected `OffsetArrays`
└ @ Requires ~/.julia/packages/ArrayInterface/mI7ab/src/ArrayInterface.jl:1116
┌ Debug: Requires conditionally ran code in 0.022674606 seconds: `HDF5` detected `FileIO`
└ @ Requires ~/.julia/packages/HDF5/pIJra/src/HDF5.jl:2133
┌ Debug: Requires conditionally ran code in 0.002288416 seconds: `RandomNumbers` detected `Random123`
└ @ Requires ~/.julia/packages/RandomNumbers/3pD1N/src/RandomNumbers.jl:38
┌ Debug: Requires conditionally ran code in 1.708197842 seconds: `CUDA` detected `SpecialFunctions`
└ @ Requires ~/.julia/packages/CUDA/sCev8/src/initialization.jl:35
┌ Debug: Requires conditionally ran code in 0.0126559 seconds: `ArrayInterface` detected `Adapt`
└ @ Requires ~/.julia/packages/ArrayInterface/mI7ab/src/ArrayInterface.jl:796
┌ Debug: Requires conditionally ran code in 0.029985851 seconds: `ArrayInterface` detected `CUDA`
└ @ Requires ~/.julia/packages/ArrayInterface/mI7ab/src/ArrayInterface.jl:795
┌ Debug: Requires conditionally ran code in 2.06503522 seconds: `Zygote` detected `CUDA`
└ @ Requires ~/.julia/packages/Zygote/umM0L/src/lib/broadcast.jl:251
┌ Debug: Requires conditionally ran code in 0.211960078 seconds: `Zygote` detected `Distances`
└ @ Requires ~/.julia/packages/Zygote/umM0L/src/Zygote.jl:45
┌ Debug: Requires conditionally ran code in 0.258361828 seconds: `Zygote` detected `LogExpFunctions`
└ @ Requires ~/.julia/packages/Zygote/umM0L/src/Zygote.jl:46
┌ Debug: Requires conditionally ran code in 0.003038184 seconds: `Zygote` detected `Colors`
└ @ Requires ~/.julia/packages/Zygote/umM0L/src/Zygote.jl:60
┌ Debug: Requires conditionally ran code in 0.284711345 seconds: `ImageMorphology` detected `ImageMetadata`
└ @ Requires ~/.julia/packages/ImageMorphology/6oqcS/src/ImageMorphology.jl:65
┌ Debug: Requires conditionally ran code in 0.002487405 seconds: `RegisterCore` detected `ImageMetadata`
└ @ Requires ~/.julia/packages/RegisterCore/YuSsR/src/RegisterCore.jl:397

julia> 
```

cc. @ChrisRackauckas I think you were also after something like this

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Beautiful!

**Comment by KristofferC** [comment]
Nice, looks like a good idea to me at least.

**Comment by ChrisRackauckas** [comment]
Yeah I don't know how to understand this. When I run it, I get:

```julia
using Profile
ENV["JULIA_DEBUG"]="Requires"
@time @profile using DifferentialEquations 

┌ Debug: Requires conditionally ran code in 0.0024978 seconds: `ArrayInterface` detected `SuiteSparse`
└ @ Requires C:\Users\accou\.julia\packages\ArrayInterface\TCj9U\src\ArrayInterface.jl:677
┌ Debug: Requires conditionally ran code in 0.0158724 seconds: `ArrayInterface` detected `StaticArrays`
└ @ Requires C:\Users\accou\.julia\packages\ArrayInterface\TCj9U\src\ArrayInterface.jl:692
┌ Debug: Requires conditionally ran code in 0.0029026 seconds: `ArrayInterface` detected `LabelledArrays`        
└ @ Requires C:\Users\accou\.julia\packages\ArrayInterface\TCj9U\src\ArrayInterface.jl:772
┌ Debug: Requires conditionally ran code in 0.0025199 seconds: `ArrayInterface` detected `Adapt`
└ @ Requires C:\Users\accou\.julia\packages\ArrayInterface\TCj9U\src\ArrayInterface.jl:762
┌ Debug: Requires conditionally ran code in 0.0168244 seconds: `Polyester` detected `ForwardDiff`
└ @ Requires C:\Users\accou\.julia\packages\Polyester\hFLHn\src\Polyester.jl:19
┌ Debug: Requires conditionally ran code in 0.0026288 seconds: `Setfield` detected `StaticArrays`
└ @ Requires C:\Users\accou\.julia\packages\Setfield\9IaGB\src\Setfield.jl:31
┌ Debug: Requires conditionally ran code in 0.0060201 seconds: `ArrayInterface` detected `OffsetArrays`
└ @ Requires C:\Users\accou\.julia\packages\ArrayInterface\TCj9U\src\ArrayInterface.jl:914
┌ Debug: Requires conditionally ran code in 0.0020892 seconds: `RandomNumbers` detected `Random123`
└ @ Requires C:\Users\accou\.julia\packages\RandomNumbers\3pD1N\src\RandomNumbers.jl:38
┌ Debug: Requires conditionally ran code in 0.0042834 seconds: `ArrayInterface` detected `BandedMatrices`        
└ @ Requires C:\Users\accou\.julia\packages\ArrayInterface\TCj9U\src\ArrayInterface.jl:809
  6.158199 seconds (19.59 M allocations: 1.412 GiB, 4.35% gc time, 14.08% compilation time)

using ProfileView
ProfileView.view()
```

Screenshot 2022-01-11 143326

And this has puzzled me for weeks. I think the debug timings are correct because I mentioned to @KristofferC a few weeks ago that I went through DiffEq's stack and just removed all of the Requires usage from almost all downstream packages and saw almost no compile-time reduction, and those timings of about a 0.02 change match what I saw. So then it really confuses me that it takes 6 seconds and the profile says almost all of those 6 seconds are in Requires.jl. Either the profile is wrong or Requires.jl is doing something really horrifying even when it's not used to init.

**Comment by vchuravy** [comment]
What is `ProfileView.view(;C=true)`?

**Comment by ChrisRackauckas** [comment]
A nice cityscape when viewed from a distance.

Screenshot 2022-01-11 144711

Including the profile:

[diffeq_requires_profile_with_C.zip](https://github.com/JuliaPackaging/Requires.jl/files/7849517/diffeq_requires_profile_with_C.zip)

though it's really easy to recreate it if you need it locally.

**Comment by IanButterworth** [comment]
I just ran this myself. The orange bars are `loading.jl, _include_from_serialized` for me. I can't see `Requires` in any of the big blocks? Am I missing something, or am I getting different behavior?

**Comment by ChrisRackauckas** [comment]
OOOOOOOOOOOOOOOOOOOOOOOOOOOOH, loading.jl require, not Requires 😅 . Carry on. This debug information looks to match the profile then where I see little ArrayInterface.jl mountains which are ultimately insignificant. Awesome!

**Comment by IanButterworth** [comment]
haha! Great.
On mine, I've only found Requires in the section I'm hovering over here

Screenshot from 2022-01-11 15-01-14

**Comment by IanButterworth** [comment]
Tim's handled the last few merges and releases here, but seems pretty busy. Does anyone else have merge & release rights?

File: src/Requires.jl [hunk 1]
```diff
@@ -54,7 +54,8 @@ end
 if isprecompiling()
     precompile(loadpkg, (Base.PkgId,)) || @warn "Requires failed to precompile `loadpkg`"
     precompile(withpath, (Any, String)) || @warn "Requires failed to precompile `withpath`"
-    precompile(err, (Any, Module, String)) || @warn "Requires failed to precompile `err`"
+    precompile(err, (Any, Module, String, String, Int)) || @warn "Requires failed to precompile `err`"
+    precompile(err, (Any, Module, String, String, Nothing)) || @warn "Requires failed to precompile `err`"
```
Comment (timholy) [inline src/Requires.jl:58]: We could clear out all the warnings (now that Julia 1.7 does this automatically), but that could be a separate PR.

**Review by timholy — APPROVED** [review]

**Comment by timholy** [comment]
Thanks! https://github.com/JuliaRegistries/General/pull/52183

> Tim's handled the last few merges and releases here, but seems pretty busy

Yup. Race against time to make progress on latency before the semester starts in earnest...
